keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,437 @@
+  different purpose.
+
+## Proposal
+
+This proposal aims to support running pods inside user namespaces. This will
+improve the pod to node isolation (phase 1 and 2) and pod to pod isolation
+(phase 3) we currently have.
+
+This mitigates all the vulnerabilities listed in the motivation section.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privileges on the host or other pods.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+#### Story 3
+
+As a cluster admin, I want to allow users to run their container as root
+without that process having root privileged on the host, so I can mitigate the
+impact of a compromised container.
+
+#### Story 4
+
+As a cluster admin, I want to allow users to choose the user/group they use
+inside the container without it being the user/group they use on the node, so I
+can mitigate the impact of a container breakout vulnerability (e.g to read
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+This field will be used for phase 1, 2 and 3.
+
+- `pod.spec.securityContex.userns.pod2podIsolation`: Enum
+If enabled, we will make the userns mappings be non-overlapping as much as possible.
+This field will be used in phase 3.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
```

Comment:
the explicit CSI field seems like a good idea.

@rata what do you think?

Reply:
@ddebroy great idea, thanks!

The ephemeral local CSI volumes might be a great fit for more isolation. But a CSI field along the lines of what exactly you mean? fsGroup? Not sure I follow :-)

I do want phase 1 to be as simple as possible, and to improve incrementally on that work. This high-level overview might seem simple, but there are _lot_ of other things we need to do behind the scenes (will add more details in follow-up PRs) for this to work. See [this comment][link] where I say in some part some of the things behind the scenes (there are way more, though :/) just to be sure we are all in the same page.

[link]: https://github.com/kubernetes/enhancements/pull/2101#issuecomment-957696725

I think this might need what we build in phase 1, maybe what we build in phase 2 too. So, I wouldn't add this to phase 1, but maybe to phase1.5/2.5 that can be done in parallel (when we look at the fine print, we will see what plumbing this really needs and better know if it can be before phase 2 or after).

However, I don't think the new PVC access mode will work with non-overlapping mappings. As long as the storage is persisted longer than the pod, it is game over, you need to have the same mappings every time you will read those files. So, this can't work with non-overlapping mappings, because:
 * You create a pod, is using mapping A. This pod uses a volume with this PVC mode and writes some files
 * This pod is destroyed and a new one is created and attached that very same volume. This pod uses mapping B, non overlap with A
 * This pod can't read the files, as they were created using a different effective UID/GID

I guess in the case of ephemeral local CSI volumes, there is no way to persist files that live longer than the pod, right? If not, we will have issues too.

It is also worth noting that as long as the volumes are ephemeral, the pods _can_ use the same mapping and be moved to non-overlapping mappings as we incrementally improve the implementation. The user should not notice this. So, I'd _really_ like to make small steps and improve incrementally, instead of making each step/phase bigger.

Will you be able/willing to work on this for this KEP? That will be great :)

Reply:
> Will you be able/willing to work on this for this KEP? That will be great :)

Yes, I would be happy to contribute from the sig-storage side to this! We are really interested in exploring alternatives to the existing fsGroup related recursive chown-ing (like the new idmaps support in the kernel) although I understand that does not fall in the scope of the immediate phase one.

> As long as the storage is persisted longer than the pod

You are correct - the lifecycle of a PV bound to a PVC with ReadWriteOncePod is indeed independent of the lifecycle of pods that mount the PVC. So the scenario you mentioned would lead to the latter pod being unable to perform IO. Will think more if associating a Group ID with a PVC and using that Group ID in the namespace mappings for all pods that want to use the PVC helps.

> But a CSI field along the lines of what exactly you mean?

I was thinking of a field/rpc to explicitly report whether a CSI volume is ephemeral or not.

Reply:
> Yes, I would be happy to contribute from the sig-storage side to this! We are really interested in exploring alternatives to the existing fsGroup related recursive chown-ing (like the new idmaps support in the kernel) although I understand that does not fall in the scope of the immediate phase one.

Yeah, I think this nicely fits into phase 3. There are two cases that might be possible here from what you mentioned, IIUC, then:
 * When we achieve pod 2 pod isolation by using the ideas already in phase 3 (per-ns/per-sa or id mapped mounts)
 * When we achieve pod 2 pod isolation by making the mappings for some cases to be non-overlapping (like ephemeral local CSI volumes).

The first case should probably work with the already planned work, the second case will need some extra work to make sure the effective UID (euid) is used for these volumes, maybe. This might not be trivial to get, I really don't know much about CSI. So I greatly appreciate your input in all, but specially on CSI =)

I'll add a reference for the second case in phase 3 (the first one should already be covered). Let me know what you think. Do you mind if I use your github handle there to ping you for this part? :)

Reply:
@ddebroy Added a mention to this into phase 3 and, hope it is okay, mentioned that you offered to help!

I think this is solved now, but please re-open if it is not! :)